config-object-store/src/main/java/org/hypertrace/config/objectstore/IdentifiedFilterPushedDownObjectStore.java

```diff
@@ -0,0 +1,64 @@
+package org.hypertrace.config.objectstore;
+
+import io.grpc.Status;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.hypertrace.config.service.change.event.api.ConfigChangeEventGenerator;
+import org.hypertrace.config.service.v1.ConfigServiceGrpc;
+import org.hypertrace.config.service.v1.Filter;
+import org.hypertrace.config.service.v1.Pagination;
+import org.hypertrace.config.service.v1.SortBy;
+import org.hypertrace.core.grpcutils.context.RequestContext;
+
+public abstract class IdentifiedFilterPushedDownObjectStore<T, F, S>
+    extends IdentifiedObjectStore<T> {
+
+  protected IdentifiedFilterPushedDownObjectStore(
+      ConfigServiceGrpc.ConfigServiceBlockingStub configServiceBlockingStub,
+      String resourceNamespace,
+      String resourceName,
+      ConfigChangeEventGenerator configChangeEventGenerator,
+      ClientConfig clientConfig) {
+    super(
+        configServiceBlockingStub,
+        resourceNamespace,
+        resourceName,
+        configChangeEventGenerator,
+        clientConfig);
+  }
+
+  public List<ContextualConfigObject<T>> getMatchingObjects(
+      RequestContext context, F filterInput, List<S> sortInput, Pagination pagination) {
+    Filter filter = buildFilter(filterInput);
+    List<SortBy> sortByList = sortInput.stream().map(this::buildSort).collect(Collectors.toList());
+    return getMatchingObjectsWithFilter(context, filter, pagination, sortByList);
+  }
+
+  public List<T> getMatchingData(
+      RequestContext context, F filterInput, List<S> sortInput, Pagination pagination) {
+    return getMatchingObjects(context, filterInput, sortInput, pagination).stream()
+        .map(ConfigObject::getData)
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  public Optional<ContextualConfigObject<T>> getMatchingObject(
+      RequestContext context, F filterInput, List<S> sortInput) {
+    List<ContextualConfigObject<T>> results =
+        getMatchingObjects(context, filterInput, sortInput, Pagination.getDefaultInstance());
```

Comment:
as soon as you see code with `getDefaultInstance()` we should be thinking "antipattern!". If we don't want pagination, don't use it. Explicitly assigning a default value to indicate "unset" breaks presence checking and even if it works today that relies on how downstream code is written which makes it really easy to introduce bugs later since you're breaking the protobuf contract.

Reply:
Updated